Rebel Binary & BuckleScript support

vramana opened the pull request:
Objectives:
- [x]  Support BuckleScript
- [x]  Build rebel with rules staticly built into jenga
- [x] Builds using npm packages
- [x]  Handle Interface files
- [ ]  Move merlin generation to seperate command
- [x]  Refactor and Comments
- [x]  Update Readme.md
- [x]  Examples
- [x]  Multiple Entry Points
- [x] Support core with native backend

vramana committed:
[WIP] Statically link rebel to jenga

Temporarily uses oasis to build rebel

vramana committed:
[WIP] Compile with npm-opam packages

vramana committed:
Use yojson in rebel.re to replace all the buildUtils

:tada: :tada:

vramana committed:
:fire: :fire: Remove unnecessary Dep based logic

vramana committed:
Change filenames

vramana committed:
Remove compile and run

vramana commented on src/rebel.re:
```diff
@@ -0,0 +1,497 @@
+/*
+  =====================================================
+    COPIED CODE BEGIN
+  =====================================================
```
All of the Coped code can be removed when Reason supports 4.03 and npm-opam/jenga use 114 major

vramana committed:
Get top level name from package.json

vramana committed:
Clean up unnecessary Dep.return

vramana committed:
Condition compilation via rebel.backend

vramana committed:
Fix Conditional compilation for jsoo

vramana committed:
Add rebel-project example and rebel.sh

vramana committed:
Add a bucklescript

vramana committed:
Commit oasis files

vramana committed:
Ignore node_modules and examples in ocamlbuild

vramana committed:
Seperate common stuff into utils.re

vramana committed:
Rebel compiles Buckscript project with only src/Index.re

vramana committed:
Compiles bs-project without any dependencies

vramana committed:
Further modularisation

Now Native and Bucklescript schemes are fully decoupled and common
parts are push into `npmDep.re` and `utils.re`

setup.re is the common entry point

vramana committed:
rebel now supports bucklescript backend with dependencies :tada: :tada:

vramana committed:
Fix syntax highlighting in github

vramana committed:
Fix script error

vramana committed:
Dependencies are namespaced with bucklescript backend

vramana committed:
Added workaround for bucklescript#757

We copy the generated file to the require calls expected path

vramana committed:
convert the case of files in bs-project to avoid bsc limitation

vramana committed:
Rebel Binary now builds from npm packages only

vramana committed:
Bye bye oasis

vramana committed:
BuckleScript scheme handles ocamlFindDeps

vramana committed:
open Utils works. Remove all the verbosity

vramana committed:
rei files work with rebel (both native and bucklescript)

Also short update README

vramana committed:
Refactor bucklescript.re and native.re + comments

vramana committed:
Reorganised stuff in utils.re and toplevel builds at _build/src

vramana committed:
rei file to reason-project example

vramana commented on src/utils.re:
```diff
@@ -0,0 +1,158 @@
+    fun name => Lib name
+  )
+}; */
+let topLibName = Lib "src";
```
@chenglou What should be the top level build dir name `src` or `name` from package.json?

chenglou replied:
You mean src name? Default to src. We can make it configurable but that's not needed right now right?

vramana replied:
Yeah. All the top level artificats will go to `_build/src`  in this implemenation.

vramana committed:
Minor refactor

vramana committed:
handle ocamlfind deps in bsc

vramana committed:
bucklescript support for multiple targets & new fields

bscFlags
ocamlcFlags
targets
backends => backend

vramana committed:
Refactor to support multiple entry points for jsoo

Disabled module aliased on the top level module
and bringing the native implementation identical to bucklescript

vramana committed:
Handle case when no targets are specified for bucklescript backend

vramana committed:
Core works with native backend

vramana committed:
Rebel bytecode :tada: :tada:

vramana committed:
bucklescript backend also generates merlin files

merlin scheme is seperated into it's own file

vramana commented on src/utils.re:
```diff
@@ -0,0 +1,195 @@
+  }
+};
+
+let bscFlags = {
```
We are not using this yet. May be I'll be scrap this future.

vramana committed:
minor edits

vramana committed:
Add -custom flag to ocamlc to avoid byte not find .so libraries

vramana committed:
remove the ocamlopt script

vramana committed:
minor edits

vramana committed:
change the rebel to install from github repo

vramana committed:
Add bsppx to merlin if the backend is bucklescript

vramana commented on src/utils.re:
```diff
@@ -0,0 +1,193 @@
+  }
+};
+
+let ocamlcFlags = {
```
@chenglou  This can handle #12 I didn't completely flesh it. Since there is still some time before this gets merged. Let me complete this also.

vramana committed:
Fix postintall

vramana committed:
don't add -open to merlin flags in the top level .merlin

vramana committed:
Remove merlin generation from build process

vramana committed:
remove the sneaky rebel.native binary

vramana committed:
fix package git urls

vramana committed:
minor fixes

vramana committed:
Revert cmi related changes

chenglou commented:
Tested. Extensively discussed in chat room. Merging, thanks!
